File: backend/app/core/pipeline_store.py
from typing import Dict, Optional
from app.models.pipeline import PipelineConfig, PipelineLayer
from app.models.node import NodeConfig, NodeRole


class PipelineStore:
    """In-memory store for pipeline configurations."""

    def __init__(self):
        self._pipelines: Dict[str, PipelineConfig] = {}
        self._init_default_pipeline()

    def _init_default_pipeline(self):
        """Initialize the default 3-layer consensus pipeline."""
        default = PipelineConfig(
            id="default",
            name="Default 3-Layer Consensus",
            description="3 generators → 2 aggregators → 1 final decision maker",
            layers=[
                PipelineLayer(
                    level=0,
                    nodes=[
                        NodeConfig(
                            id="gen-1",
                            provider="openai",
                            model="gpt-4",
                            role=NodeRole.GENERATOR,
                        ),
                        NodeConfig(
                            id="gen-2",
                            provider="anthropic",
                            model="claude-3-sonnet-20240229",
                            role=NodeRole.GENERATOR,
                        ),
                        NodeConfig(
                            id="gen-3",
                            provider="mistral",
                            model="mistral-large-latest",
                            role=NodeRole.GENERATOR,
                        ),
                    ],
                ),
                PipelineLayer(
                    level=1,
                    nodes=[
                        NodeConfig(
                            id="agg-1",
                            provider="openai",
                            model="gpt-4",
                            role=NodeRole.AGGREGATOR,
                            system_prompt="You are an aggregator. Analyze the following responses and synthesize the key points into a coherent summary.",
                        ),
                        NodeConfig(
                            id="agg-2",
                            provider="anthropic",
                            model="claude-3-sonnet-20240229",
                            role=NodeRole.AGGREGATOR,
                            system_prompt="You are an aggregator. Analyze the following responses and synthesize the key points into a coherent summary.",
                        ),
                    ],
                ),
                PipelineLayer(
                    level=2,
                    nodes=[
                        NodeConfig(
                            id="final-1",
                            provider="openai",
                            model="gpt-4",
                            role=NodeRole.FINAL,
                            system_prompt="You are the final decision maker. Based on the aggregated responses, provide the best possible answer.",
                        ),
                    ],
                ),
            ],
        )
        self._pipelines["default"] = default

    def get(self, pipeline_id: str) -> Optional[PipelineConfig]:
        return self._pipelines.get(pipeline_id)

    def get_all(self) -> Dict[str, PipelineConfig]:
        return self._pipelines.copy()

    def save(self, config: PipelineConfig) -> None:
        self._pipelines[config.id] = config

    def delete(self, pipeline_id: str) -> bool:
        if pipeline_id in self._pipelines:
            del self._pipelines[pipeline_id]
            return True
        return False


# Global singleton
pipeline_store = PipelineStore()

File: frontend/src/stores/pipelineStore.ts
import { create } from 'zustand';
import { PipelineConfig, PipelineState, NodeState, NodeStatus } from '../types/pipeline';

interface PipelineStore {
  // Current pipeline config
  currentPipeline: PipelineConfig | null;
  pipelines: PipelineConfig[];

  // Execution state
  executionState: PipelineState | null;
  isExecuting: boolean;

  // Actions
  setCurrentPipeline: (pipeline: PipelineConfig | null) => void;
  setPipelines: (pipelines: PipelineConfig[]) => void;
  addPipeline: (pipeline: PipelineConfig) => void;
  removePipeline: (id: string) => void;

  // Execution actions
  startExecution: (executionId: string, pipelineId: string) => void;
  updateNodeState: (nodeId: string, state: Partial<NodeState>) => void;
  updateExecutionProgress: (currentLayer: number, progress: number) => void;
  finishExecution: (finalOutput: string, consensusScore: number) => void;
  resetExecution: () => void;
}

const defaultPipeline: PipelineConfig = {
  id: 'default',
  name: 'Default 3-Layer Consensus',
  description: '3 generators → 2 aggregators → 1 final',
  layers: [
    {
      level: 0,
      nodes: [
        { id: 'gen-1', provider: 'openai', model: 'gpt-4', role: 'generator' },
        { id: 'gen-2', provider: 'anthropic', model: 'claude-3-sonnet', role: 'generator' },
        { id: 'gen-3', provider: 'mistral', model: 'mistral-large', role: 'generator' },
      ],
    },
    {
      level: 1,
      nodes: [
        { id: 'agg-1', provider: 'openai', model: 'gpt-4', role: 'aggregator' },
        { id: 'agg-2', provider: 'anthropic', model: 'claude-3-sonnet', role: 'aggregator' },
      ],
    },
    {
      level: 2,
      nodes: [
        { id: 'final-1', provider: 'openai', model: 'gpt-4', role: 'final' },
      ],
    },
  ],
};

export const usePipelineStore = create<PipelineStore>((set) => ({
  currentPipeline: defaultPipeline,
  pipelines: [defaultPipeline],
  executionState: null,
  isExecuting: false,

  setCurrentPipeline: (pipeline) => set({ currentPipeline: pipeline }),

  setPipelines: (pipelines) => set({ pipelines }),

  addPipeline: (pipeline) =>
    set((state) => ({
      pipelines: [...state.pipelines, pipeline],
    })),

  removePipeline: (id) =>
    set((state) => ({
      pipelines: state.pipelines.filter((p) => p.id !== id),
    })),

  startExecution: (executionId, pipelineId) =>
    set((state) => {
      const pipeline = state.pipelines.find((p) => p.id === pipelineId);
      if (!pipeline) return state;

      const nodeStates: Record<string, NodeState> = {};
      pipeline.layers.forEach((layer) => {
        layer.nodes.forEach((node) => {
          nodeStates[node.id] = {
            nodeId: node.id,
            status: 'pending' as NodeStatus,
          };
        });
      });

      return {
        isExecuting: true,
        executionState: {
          executionId,
          pipelineId,
          status: 'running',
          currentLayer: 0,
          nodeStates,
        },
      };
    }),

  updateNodeState: (nodeId, state) =>
    set((prev) => {
      if (!prev.executionState) return prev;

      return {
        executionState: {
          ...prev.executionState,
          nodeStates: {
            ...prev.executionState.nodeStates,
            [nodeId]: {
              ...prev.executionState.nodeStates[nodeId],
              ...state,
            },
          },
        },
      };
    }),

  updateExecutionProgress: (currentLayer, progress) =>
    set((state) => {
      if (!state.executionState) return state;

      return {
        executionState: {
          ...state.executionState,
          currentLayer,
        },
      };
    }),

  finishExecution: (finalOutput, consensusScore) =>
    set((state) => {
      if (!state.executionState) return state;

      return {
        isExecuting: false,
        executionState: {
          ...state.executionState,
          status: 'completed',
          finalOutput,
          consensusScore,
        },
      };
    }),

  resetExecution: () =>
    set({
      isExecuting: false,
      executionState: null,
    }),
}));
